Userland/Libraries/LibVirtGPU/Device.cpp

```diff
@@ -0,0 +1,449 @@
+        .width = static_cast<u32>(size.width()),
+        .height = static_cast<u32>(size.height()),
+        .depth = 1,
+        .array_size = 1,
+        .last_level = 0,
+        .nr_samples = 0,
+        .flags = 0,
+        .created_resource_id = 0,
+    };
+    m_depthbuffer_surface = create_virgl_resource(depthbuffer_surface_spec);
+
+    // Initialize all required state
+    CommandBufferBuilder builder;
+
+    // Create and set the blend, to control the color mask
+    m_blend_handle = allocate_handle();
+    builder.append_create_blend(m_blend_handle);
+    builder.append_bind_blend(m_blend_handle);
+
+    // Create drawtarget surface
+    m_drawtarget_surface_handle = allocate_handle();
+    builder.append_create_surface(m_drawtarget, m_drawtarget_surface_handle, Protocol::TextureFormat::VIRTIO_GPU_FORMAT_B8G8R8A8_UNORM);
+
+    // Create depthbuffer surface
+    m_depthbuffer_surface_handle = allocate_handle();
+    builder.append_create_surface(m_depthbuffer_surface, m_depthbuffer_surface_handle, Protocol::TextureFormat::VIRTIO_GPU_FORMAT_Z32_FLOAT);
+
+    // Set some framebuffer state (attached handle, framebuffer size, etc)
+    builder.append_set_framebuffer_state(m_drawtarget_surface_handle, m_depthbuffer_surface_handle);
+    builder.append_set_framebuffer_state_no_attach(size);
+
+    // Set the vertex buffer
+    builder.append_set_vertex_buffers(sizeof(VertexData), 0, m_vbo_resource_id);
+
+    // Create and bind fragment shader
+    m_frag_shader_handle = allocate_handle();
+    builder.append_create_shader(m_frag_shader_handle, Gallium::ShaderType::SHADER_FRAGMENT, frag_shader);
+    builder.append_bind_shader(m_frag_shader_handle, Gallium::ShaderType::SHADER_FRAGMENT);
+
+    // Create and bind vertex shader
+    m_vert_shader_handle = allocate_handle();
+    builder.append_create_shader(m_vert_shader_handle, Gallium::ShaderType::SHADER_VERTEX, vert_shader);
+    builder.append_bind_shader(m_vert_shader_handle, Gallium::ShaderType::SHADER_VERTEX);
+
+    // Create a VertexElements object (used to specify layout of vertex data)
+    m_ve_handle = allocate_handle();
+    builder.append_create_vertex_elements(m_ve_handle);
+    builder.append_bind_vertex_elements(m_ve_handle);
+
+    // Create a DepthStencilAlpha (DSA) object
+    m_dsa_handle = allocate_handle();
+    builder.append_create_dsa(m_dsa_handle);
+    builder.append_bind_dsa(m_dsa_handle);
+
+    // Create a Rasterizer object
+    m_rasterizer_handle = allocate_handle();
+    builder.append_create_rasterizer(m_rasterizer_handle);
+    builder.append_bind_rasterizer(m_rasterizer_handle);
+
+    // Set the Viewport
+    builder.append_gl_viewport(size);
+
+    // Upload buffer
+    upload_command_buffer(builder.build());
+}
+
+Device::~Device()
+{
+    close(m_gpu_file_descriptor);
+}
+
+GPU::DeviceInfo Device::info() const
+{
+    return {
+        .vendor_name = "SerenityOS",
+        .device_name = "VirtIO-GPU",
```

Comment:
`LibSoftGPU` -> `SoftGPU`, maybe call this `VirtGPU`? Or the library `LibVirtIOGPU`?